// @nguniversal/common/bundles/common.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common/http'), require('@angular/core'), require('@angular/platform-browser'), require('rxjs/observable/of'), require('rxjs/operators/filter'), require('rxjs/operators/tap'), require('rxjs/operators/take')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/common/http', '@angular/core', '@angular/platform-browser', 'rxjs/observable/of', 'rxjs/operators/filter', 'rxjs/operators/tap', 'rxjs/operators/take'], factory) :
	(factory((global.nguniversal = global.nguniversal || {}, global.nguniversal.common = global.nguniversal.common || {}),global.ng.common.http,global.ng.core,global.ng.platformBrowser,global.Rx.Observable,global.Rx.operators,global.Rx.operators,global.Rx.operators));
}(this, (function (exports,_angular_common_http,_angular_core,_angular_platformBrowser,rxjs_observable_of,rxjs_operators_filter,rxjs_operators_tap,rxjs_operators_take) { 'use strict';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */

/**
 * @param {?} headers
 * @return {?}
 */
function getHeadersMap(headers) {
    var /** @type {?} */ headersMap = {};
    for (var _i = 0, _a = headers.keys(); _i < _a.length; _i++) {
        var key = _a[_i];
        headersMap[key] = /** @type {?} */ ((headers.getAll(key)));
    }
    return headersMap;
}
var TransferHttpCacheInterceptor = /** @class */ (function () {
    function TransferHttpCacheInterceptor(appRef, transferState) {
        var _this = this;
        this.transferState = transferState;
        this.isCacheActive = true;
        // Stop using the cache if the application has stabilized, indicating initial rendering is
        // complete.
        appRef.isStable
            .pipe(rxjs_operators_filter.filter(function (isStable) { return isStable; }), rxjs_operators_take.take(1)).toPromise()
            .then(function () { _this.isCacheActive = false; });
    }
    /**
     * @param {?} url
     * @return {?}
     */
    TransferHttpCacheInterceptor.prototype.invalidateCacheEntry = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        this.transferState.remove(_angular_platformBrowser.makeStateKey('G.' + url));
        this.transferState.remove(_angular_platformBrowser.makeStateKey('H.' + url));
    };
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    TransferHttpCacheInterceptor.prototype.intercept = /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    function (req, next) {
        var _this = this;
        // Stop using the cache if there is a mutating call.
        if (req.method !== 'GET' && req.method !== 'HEAD') {
            this.isCacheActive = false;
            this.invalidateCacheEntry(req.url);
        }
        if (!this.isCacheActive) {
            // Cache is no longer active. Pass the request through.
            return next.handle(req);
        }
        var /** @type {?} */ key = (req.method === 'GET' ? 'G.' : 'H.') + req.url;
        var /** @type {?} */ storeKey = _angular_platformBrowser.makeStateKey(key);
        if (this.transferState.hasKey(storeKey)) {
            // Request found in cache. Respond using it.
            var /** @type {?} */ response = this.transferState.get(storeKey, /** @type {?} */ ({}));
            return rxjs_observable_of.of(new _angular_common_http.HttpResponse({
                body: response.body,
                headers: new _angular_common_http.HttpHeaders(response.headers),
                status: response.status,
                statusText: response.statusText,
                url: response.url,
            }));
        }
        else {
            // Request not found in cache. Make the request and cache it.
            var /** @type {?} */ httpEvent = next.handle(req);
            return httpEvent
                .pipe(rxjs_operators_tap.tap(function (event) {
                if (event instanceof _angular_common_http.HttpResponse) {
                    _this.transferState.set(storeKey, {
                        body: event.body,
                        headers: getHeadersMap(event.headers),
                        status: event.status,
                        statusText: event.statusText,
                        url: /** @type {?} */ ((event.url)),
                    });
                }
            }));
        }
    };
    TransferHttpCacheInterceptor.decorators = [
        { type: _angular_core.Injectable },
    ];
    /** @nocollapse */
    TransferHttpCacheInterceptor.ctorParameters = function () { return [
        { type: _angular_core.ApplicationRef, },
        { type: _angular_platformBrowser.TransferState, },
    ]; };
    return TransferHttpCacheInterceptor;
}());
/**
 * An NgModule used in conjunction with `ServerTransferHttpCacheModule` to transfer cached HTTP
 * calls from the server to the client application.
 */
var TransferHttpCacheModule = /** @class */ (function () {
    function TransferHttpCacheModule() {
    }
    TransferHttpCacheModule.decorators = [
        { type: _angular_core.NgModule, args: [{
                    imports: [_angular_platformBrowser.BrowserTransferStateModule],
                    providers: [
                        TransferHttpCacheInterceptor,
                        { provide: _angular_common_http.HTTP_INTERCEPTORS, useExisting: TransferHttpCacheInterceptor, multi: true },
                    ],
                },] },
    ];
    /** @nocollapse */
    TransferHttpCacheModule.ctorParameters = function () { return []; };
    return TransferHttpCacheModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */

exports.TransferHttpCacheModule = TransferHttpCacheModule;
exports.ɵa = TransferHttpCacheInterceptor;

Object.defineProperty(exports, '__esModule', { value: true });

})));
